### ISSUE TITLE
System.InvalidOperationException: Cross-thread operation not valid

### ISSUE THREAD

**Issue by kemalatli** [eXpandFramework/eXpand#553]
**Describe the bug**
After mapping properties with column names and hitting import, I get cross thread operation exception. I think there might be an issue with the subscription handling of RX library used in Excel Importer. It points to progressbar update. I'm using version 19.1.503

**To Reproduce**
1. Map columns
2. Click on import

**Expected behavior**
Successfully import all validated data.

**Additional context**

> System.InvalidOperationException
  HResult=0x80131509
  Message=Cross-thread operation not valid: Control 'ProgressBarViewItem(30)' accessed from a thread other than the thread it was created on.
  Source=System.Windows.Forms
  StackTrace:
   at System.Windows.Forms.Control.get_Handle()
   at System.Windows.Forms.Control.PointToClientInternal(Point p)
   at DevExpress.XtraEditors.BaseControl.LayoutChangedCore(Boolean invalidateVisual)
   at DevExpress.XtraEditors.BaseEdit.LayoutChanged()
   at DevExpress.XtraEditors.BaseEdit.OnEditValueChanged()
   at DevExpress.XtraEditors.BaseEdit.OnEditValueChanging(ChangingEventArgs e)
   at DevExpress.XtraEditors.BaseEdit.set_EditValue(Object value)
   at DevExpress.XtraEditors.ProgressBarControl.set_EditValue(Object value)
   at DevExpress.XtraEditors.ProgressBarControl.set_Position(Int32 value)
   at Fasterflect.PropertyExtensions.SetPropertyValue(Object obj, String name, Object value)
   at Xpand.XAF.Modules.ProgressBarViewItem.ProgressBarViewItemBase.SetPosition(Decimal value)
   at Xpand.XAF.Modules.ProgressBarViewItem.ProgressBarViewItemBase.<Start>b__14_0()
   at System.Reactive.Linq.ObservableImpl.Finally`1._.Dispose(Boolean disposing) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Finally.cs:line 75
   at System.Reactive.Sink`1.Dispose() in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\Sink.cs:line 31
   at System.Reactive.Disposables.Disposable.TryDispose(IDisposable& fieldRef) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Disposables\Disposable.cs:line 237
   at System.Reactive.Linq.ObservableImpl.TakeUntil`2._.Dispose(Boolean disposing) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\TakeUntil.cs:line 53
   at System.Reactive.Sink`1.Dispose() in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\Sink.cs:line 31
   at System.Reactive.Disposables.Disposable.TryDispose(IDisposable& fieldRef) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Disposables\Disposable.cs:line 237
   at System.Reactive.SafeObserver`1.Dispose(Boolean disposing) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\SafeObserver.cs:line 93
   at System.Reactive.SafeObserver`1.Dispose() in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\SafeObserver.cs:line 85
   at System.Reactive.AnonymousSafeObserver`1.OnError(Exception error) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\AnonymousSafeObserver.cs:line 65
   at System.Reactive.Sink`1.ForwardOnError(Exception error) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\Sink.cs:line 62
   at System.Reactive.HalfSerializer.ForwardOnError[T](ISink`1 sink, Exception ex, Int32& wip, Exception& error) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\HalfSerializer.cs:line 79
   at System.Reactive.Linq.ObservableImpl.TakeUntil`2._.OnError(Exception ex) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\TakeUntil.cs:line 63
   at System.Reactive.Sink`2.OnError(Exception error) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\Sink.cs:line 97
   at System.Reactive.Linq.ObservableImpl.Do`1.OnNext._.OnNext(TSource value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Do.cs:line 47
   at System.Reactive.Subjects.Subject`1.OnNext(T value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Subjects\Subject.cs:line 150
   at Xpand.XAF.Modules.ProgressBarViewItem.ProgressBarViewItemBase.OnNext(Decimal value)
   at System.Reactive.Linq.ObservableImpl.Do`1.Observer._.OnNext(TSource value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Do.cs:line 89
   at System.Reactive.PlatformServices.ExceptionServicesImpl.Rethrow(Exception exception) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\ExceptionServicesImpl.cs:line 16
   at System.Reactive.Stubs.<>c.<.cctor>b__2_1(Exception ex) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\Stubs.cs:line 16
   at System.Reactive.AnonymousSafeObserver`1.OnError(Exception error) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\AnonymousSafeObserver.cs:line 65
   at System.Reactive.Linq.ObservableImpl.Do`1.Observer._.OnNext(TSource value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Do.cs:line 89
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\Sink.cs:line 50
   at System.Reactive.HalfSerializer.ForwardOnNext[T](ISink`1 sink, T item, Int32& wip, Exception& error) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\HalfSerializer.cs:line 55
   at System.Reactive.Linq.ObservableImpl.TakeUntil`2._.OnNext(TSource value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\TakeUntil.cs:line 58
   at System.Reactive.Linq.ObservableImpl.ConcatMany`1.ConcatManyOuterObserver.InnerObserver.OnNext(T value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\ConcatMany.cs:line 221
   at System.Reactive.Linq.ObservableImpl.ReturnImmediate`1.Run(IObserver`1 observer) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Return.cs:line 64
   at System.Reactive.BasicProducer`1.<>c.<SubscribeRaw>b__1_0(ValueTuple`3 tuple) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\Producer.cs:line 60
   at System.Reactive.Concurrency.Scheduler.<>c__76`1.<ScheduleAction>b__76_0(IScheduler _, ValueTuple`2 tuple) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\Scheduler.Simple.cs:line 88
   at System.Reactive.Concurrency.CurrentThreadScheduler.Schedule[TState](TState state, TimeSpan dueTime, Func`3 action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\CurrentThreadScheduler.cs:line 148
   at System.Reactive.Concurrency.LocalScheduler.Schedule[TState](TState state, Func`3 action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\LocalScheduler.cs:line 32
   at System.Reactive.Concurrency.Scheduler.ScheduleAction[TState](IScheduler scheduler, TState state, Func`2 action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\Scheduler.Simple.cs:line 86
   at System.Reactive.BasicProducer`1.SubscribeRaw(IObserver`1 observer, Boolean enableSafeguard) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\Producer.cs:line 70
   at System.ObservableExtensions.SubscribeSafe[T](IObservable`1 source, IObserver`1 observer) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Observable.Extensions.cs:line 424
   at System.Reactive.Linq.ObservableImpl.ConcatMany`1.ConcatManyOuterObserver.Drain() in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\ConcatMany.cs:line 171
   at System.Reactive.Linq.ObservableImpl.ConcatMany`1.ConcatManyOuterObserver.OnNext(IObservable`1 value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\ConcatMany.cs:line 90
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\Sink.cs:line 50
   at System.Reactive.Linq.ObservableImpl.Select`2.Selector._.OnNext(TSource value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Select.cs:line 48
   at System.Reactive.Linq.ObservableImpl.WithLatestFrom`3._.FirstObserver.OnNext(TFirst value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\WithLatestFrom.cs:line 123
   at System.Reactive.Linq.ObservableImpl.Timer.Periodic._.Tick() in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Timer.cs:line 168
   at System.Reactive.Linq.ObservableImpl.Timer.Periodic._.<>c.<Run>b__4_0(_ this) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Timer.cs:line 141
   at System.Reactive.Concurrency.Scheduler.<>c__67`1.<SchedulePeriodic>b__67_0(ValueTuple`2 t) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\Scheduler.Services.Emulation.cs:line 79
   at System.Reactive.Concurrency.DefaultScheduler.PeriodicallyScheduledWorkItem`1.<>c.<Tick>b__5_0(PeriodicallyScheduledWorkItem`1 closureWorkItem) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\DefaultScheduler.cs:line 127
   at System.Reactive.Concurrency.AsyncLock.<>c__5`1.<Wait>b__5_0(Delegate actionObject, Object stateObject) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\AsyncLock.cs:line 53
   at System.Reactive.Concurrency.AsyncLock.Wait(Object state, Delegate delegate, Action`2 action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\AsyncLock.cs:line 105
   at System.Reactive.Concurrency.AsyncLock.Wait[TState](TState state, Action`1 action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\AsyncLock.cs:line 53
   at System.Reactive.Concurrency.DefaultScheduler.PeriodicallyScheduledWorkItem`1.Tick() in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\DefaultScheduler.cs:line 125
   at System.Reactive.Concurrency.ConcurrencyAbstractionLayerImpl.PeriodicTimer.Tick(Object state) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\ConcurrencyAbstractionLayerImpl.cs:line 215
   at System.Reactive.Concurrency.ConcurrencyAbstractionLayerImpl.PeriodicTimer.<>c.<.ctor>b__2_0(Object _) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\ConcurrencyAbstractionLayerImpl.cs:line 210
   at System.Threading.TimerQueueTimer.CallCallbackInContext(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.TimerQueueTimer.CallCallback()
   at System.Threading.TimerQueueTimer.Fire()
   at System.Threading.TimerQueue.FireNextTimers()
   at System.Threading.TimerQueue.AppDomainTimerCallback(Int32 id)

**Comment by apobekiaris** [comment]
I cannot reproduce the case with the latest version, can you try the latest? If you still have issues provide sample solution and test data (xls

**Comment by apobekiaris** [comment]
oops my bad looking at the source code I found that the SynchronizationContext.Current is not used. I fixed it and the bot will notify u once ready

**Comment by kemalatli** [comment]
great to hear it is resolved! When should I expect a release?

**Comment by apobekiaris** [comment]
it won't be long the bot will update you once ready

**Comment by apobekiaris** [comment]
The [DevExpress.XAF](https://github.com/eXpandFramework/DevExpress.XAF) repository includes commits that relate to this task:

* [ProgressBarViewItem: Cross thread exception observe on SynchronizationContext.Current #553](https://github.com/eXpandFramework/DevExpress.XAF/commit/8465d426d4979a12b3e1cb380064d903ea4d3758)

Please update the related Nuget packages and test if issues is addressed. These are nightly nuget packages available only from our [NugetServer](https://xpandnugetserver.azurewebsites.net/nuget/).

If you do not use the Xpand.XAF.Modules directly but through a module of the main eXpandFramework project, please wait for the bot to notify you again when integration is finished or update the related packages manually.

Thanks a lot for your contribution.

**Comment by kemalatli** [comment]
@apobekiaris I had chance to test latest build. At first when I tested import functionality with local database, it was working. Then, I changed database connection to remote MySQL and this issue happened again.

`'Cross-thread operation not valid: Control 'ProgressBarViewItem(30)' accessed from a thread other than the thread it was created on.'`

In our web application, excel import doesn't produce this issue. It only happens with WinForms App while using MySQL connection.

**Comment by apobekiaris** [comment]
which version did u tried? there was an issue after scaling the tests to multiple agents so the working version should be

```
Pushing Xpand.XAF.Modules.ProgressBarViewItem.0.0.18.1.nupkg to '***'...
```

**Comment by apobekiaris** [comment]
note that this case has nothing to do with the datastore used

**Comment by kemalatli** [comment]
You are right! Issue resolved with version `19.1.602.7`. Thanks for the great effort on this framework!
